### ISSUE TITLE
Needleman-Wunsch matrix construction + alignment issue

### ISSUE THREAD

**Issue by notkaramel** [ECHOES-from-the-Past/mei-analyser#5]
# Issue 
When first implemented, I imagined the top row of the matrix to be the first pattern array, and the first column to be the second pattern array. However, the matrix construction function for the algorithm was flipped (or inversed) that led to a confusing indexing issue when performing the note highlighting.

# Current progress
- [x] Construction function of the matrix
- [x] "Unit" tests for each steps of the algorithm to make sure it is working as intended
- [x] Some documentation on the tests
- [x] Highlighting functions

**Comment by notkaramel** [comment]
The algorithm works now! There are some misalignment with the pattern array (input to the algorithm) that is currently being fixed. Highlighting is also shifted. I will separate the functions, since the `pattern_analysis()` function is incredibly complicated and has tons of repeated codes (which is very bad).

**Comment by martha-thomae** [comment]
Thank you! Yes, let's avoid repeated code and use function calls instead for those :)

**Comment by notkaramel** [comment]
tl;dr: the issue is fixed! 

The code is now cleaned up (a bit, I think it can still be improved). The highlighting alignment is also fixed and I'm writing some documentation about how all of these is working on [DesignDocumentation.md](https://github.com/ECHOES-from-the-Past/mei-analyser/blob/main/DesignDocumentation.md).
